File: ch01/E01G_hamming_distance.py
# Hamming Distance Problem:
# Compute the Hamming distance between two strings
#   Input: Two strings of equal length
#   Output: The Hamming distance between these strings

from utils import Text


# Complexity: O(n)
def HammingDistance(str1, str2):
    distance = 0

    for i in range(len(str1)):
        if str1[i] != str2[i]:
            distance += 1

    return distance


# Complexity:
# - For cycle: n - k
# - Calculating distance: k
# O(n - k) * O(k) = O(nk - k^2) ~ O(nk), because n >> k
def ApproximatePatternCount(text, pattern, threshold):
    count = 0
    indices = []

    for i in range(len(text) - len(pattern)):
        n_pattern = Text(text, i, len(pattern))

        if HammingDistance(pattern, n_pattern) <= threshold:
            count += 1
            indices.append("^")
        else:
            indices.append("-")

    return count, indices


if __name__ == "__main__":
    _genome = input("Genome: ").upper()
    # _genome = read_txt("../data/GCF_000813165.1_ASM81316v1_genomic_Escherichia_coli.fna", 1)

    _pattern = input("Pattern: ").upper()
    _threshold = int(input("Threshold: "))

    _count, _matches = ApproximatePatternCount(_genome, _pattern, _threshold)

    print(_count)
    print(_genome)

    for index in _matches:
        print(index, end="")

File: cpp/ch01/E01G_hamming_distance.cpp
//
// Created by Matej Hrnčiar on 21/07/2022.
//

#include "E01G_hamming_distance.h"

int HammingDistance(const string &str1, const string &str2) {
    int distance = 0;

    if (str1.size() != str2.size()) {
        return -1;
    }

    for (int i = 0; i < str1.size(); i++) {
        if (str1[i] != str2[i]) {
            distance++;
        }
    }

    return distance;
}

tuple< int, vector<char> > ApproximatePatternCount(const string &genome, const string &pattern, int threshold) {
    int count = 0;
    vector<char> indices;

    for (unsigned long i = 0; i < genome.size() - pattern.size(); i++) {
        string n_pattern = Text(genome, i, pattern.size());

        if (HammingDistance(pattern, n_pattern) <= threshold) {
            count++;
            indices.push_back('^');
        }
        else {
            indices.push_back('-');
        }
    }

    return {count, indices};
}

/*
int main() {
    string _genome, _pattern;
    int _threshold;

    cout << "Genome: ";
    cin >> _genome;
    cout << "Pattern: ";
    cin >> _pattern;
    cout << "t: ";
    cin >> _threshold;

    auto [_count, _indices] = ApproximatePatternCount(_genome, _pattern, _threshold);

    cout << "Count: " << _count << endl;
    cout << _genome << endl;

    for (char c : _indices) {
        cout << c;
    }
}
 */
